pkg/crc/machine/generate_bundle.go

```diff
@@ -0,0 +1,155 @@
+	kubePassword, err := crcBundleMetadata.GetKubeadminPassword()
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(kubeConfigAdminPasswordPath, []byte(kubePassword), 0600)
+	if err != nil {
+		return fmt.Errorf("error writing to %s: %v", kubeConfigFilePath, err)
+	}
+
+	// Copy ssh keyfile
+	sshKeyPath := filepath.Join(customBundleDir, "id_ecdsa_crc")
+	generatedKey := filepath.Join(constants.MachineInstanceDir, constants.DefaultName, "id_ecdsa")
+	if err := crcos.CopyFileContents(generatedKey, sshKeyPath, 0400); err != nil {
+		return err
+	}
+
+	// Copy the oc binary
+	ocFilePath := filepath.Join(constants.CrcOcBinDir, constants.OcExecutableName)
+	if err := crcos.CopyFileContents(ocFilePath, filepath.Join(customBundleDir, constants.OcExecutableName), 0755); err != nil {
+		return err
+	}
+
+	// Copy disk image
+	logging.Infof("Copy the disk image to %s", customBundleDir)
+	if err := copyDiskImage(customBundleDir, crcBundleMetadata); err != nil {
+		return err
+	}
+
+	// Get size of newly created disk
+	diskPath := filepath.Join(customBundleDir, filepath.Base(crcBundleMetadata.GetDiskImagePath()))
+	stat, err := os.Stat(diskPath)
+	if err != nil {
+		return err
+	}
+
+	// update bundle info
+	crcBundleMetadata.BuildInfo.BuildTime = time.Now().String()
+	crcBundleMetadata.Storage.DiskImages[0].Size = strconv.FormatInt(stat.Size(), 10)
+
+	// Create the metadata json for custom bundle
+	bundleContent, err := json.MarshalIndent(crcBundleMetadata, "", " ")
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(filepath.Join(customBundleDir, "crc-bundle-info.json"), bundleContent, 0600)
+	if err != nil {
+		return fmt.Errorf("error copying bundle metadata  %v", err)
+	}
+
+	logging.Infof("Compressing %s", customBundleDir)
+	if err := compress.Compress(customBundleDir, fmt.Sprintf("%s_custom%s", currentBundleName, constants.DefaultBundleExtension)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func removePullSecretAndStop(client *client) (*bundle.CrcBundleInfo, error) {
+	libMachineAPIClient, cleanup := createLibMachineClient()
+	defer cleanup()
+
+	host, err := libMachineAPIClient.Load(client.name)
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot load machine")
+	}
+
+	_, err = host.Driver.GetState()
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot get machine state")
+	}
+
+	crcBundleMetadata, err := getBundleMetadataFromDriver(host.Driver)
+	if err != nil {
+		return crcBundleMetadata, errors.Wrap(err, "Error loading bundle metadata")
+	}
+
+	// Remove the pull secret
+	instanceIP, err := getIP(host, client.useVSock())
+	if err != nil {
+		return crcBundleMetadata, errors.Wrap(err, "Error getting the IP")
+	}
+	sshRunner, err := crcssh.CreateRunner(instanceIP, getSSHPort(client.useVSock()), crcBundleMetadata.GetSSHKeyPath(), constants.GetPrivateKeyPath(), constants.GetRsaPrivateKeyPath())
+	if err != nil {
+		return crcBundleMetadata, errors.Wrap(err, "Error creating the ssh client")
+	}
```

Comment:
The hoops we jump through in order to get a `sshRunner` and the bundle metada still obscure what this function is doing, which starts right after this comment.
This could probably become
```
func (client *client) GenerateBundle() error {
        crcBundleMetadata, sshRunner, err := loadVM(client)
        if err != nil {
                return err
        }
        defer sshRunner.Close()
        ocConfig := oc.UseOCWithSSH(sshRunner)
        if err := cluster.RemovePullSecretFromCluster(ocConfig, sshRunner); err != nil {
                return err
        }

        currentState, err := client.Stop()
        if err != nil {
                return err
        }
        if currentState != state.Stopped {
                return fmt.Errorf("VM is not stopped, current state is %s", currentState.String())
        }

        return generateBundle(crcBundleMetadata)
}
```
which only shows the useful steps for the bundle generation. If we want more details, we can go look at the other helpers.